File: Backend/backend-api/src/user/dto/find-user-query.dto.ts
import type { IPaginationQueryDto } from '#/utils/dto/pagination.dto.js';
import typia from 'typia';
import type { IRequestQueryValidator } from '#/utils/typia/typia-type.js';
import { cleanString } from '#/utils/string.util.js';
import type { UserRole, UserStatus } from '#/generated/drizzle/enums.js';

import type { TagsSortOrder } from '#/utils/typia/validators/sort.validator.js';

export interface IFindUserQueryDto extends IPaginationQueryDto {
  search?: string;

  role?: UserRole;

  status?: UserStatus;

  selectUserStatus?: boolean;

  selectUserRole?: boolean;

  user_id?: boolean;

  username?: boolean;

  email?: boolean;

  first_name?: boolean;

  last_name?: boolean;

  telephone?: boolean;

  cif?: boolean;

  profile?: boolean;

  orderBy?:
    | 'id'
    | 'user_id'
    | 'username'
    | 'email'
    | 'first_name'
    | 'last_name'
    | 'telephone'
    | 'cif';

  orderDir?: TagsSortOrder;
}

export const validateFindUserQueryFunction =
  typia.http.createAssertQuery<IFindUserQueryDto>();

export const validateFindUserQuery: IRequestQueryValidator.IAssert<IFindUserQueryDto> =
  {
    type: 'assert',
    assert: (input): IFindUserQueryDto => {
      const search = input.get('search');
      if (search) input.set('search', cleanString(search));

      return validateFindUserQueryFunction(input);
    },
  };

File: Frontend/Maian/shared/src/commonMain/kotlin/org/dsqrwym/shared/data/user/dto/FindUserQueryDto.kt
package org.dsqrwym.shared.data.user.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.dsqrwym.shared.data.OrderDir
import org.dsqrwym.shared.paging.data.PaginationQuery
import org.dsqrwym.shared.data.user.UserRole
import org.dsqrwym.shared.data.user.UserStatus

enum class FindUserQueryOrderBy(val value: String) {
    ID("id"),
    USER_ID("user_id"),
    USERNAME("username"),

    EMAIL("email"),

    FIRST_NAME("first_name"),

    LAST_NAME("last_name"),

    TELEPHONE("telephone"),

    CIF("cif")
}

@Serializable
data class FindUserQueryDto(
    val search: String? = null,
    val role: UserRole? = null,
    val status: UserStatus? = null,
    val selectUserStatus: Boolean? = null,
    val selectUserRole: Boolean? = null,
    @SerialName("user_id")
    val userId: Boolean? = null,
    val username: Boolean? = null,
    val email: Boolean? = null,
    @SerialName("first_name")
    val firstName: Boolean? = null,
    @SerialName("last_name")
    val lastName: Boolean? = null,
    val telephone: Boolean? = null,
    val cif: Boolean? = null,
    val profile: Boolean? = null,
    val orderBy: FindUserQueryOrderBy? = null,
    val orderDir: OrderDir? = null,
    override val page: Int = 1,
    override val limit: Int = 50,
) : PaginationQuery